File: main/src/main/java/lavaplayer/source/youtube/YoutubeAudioTrack.java
package lavaplayer.source.youtube;

import lavaplayer.container.matroska.MatroskaAudioTrack;
import lavaplayer.container.mpeg.MpegAudioTrack;
import lavaplayer.source.ItemSourceManager;
import lavaplayer.tools.FriendlyException;
import lavaplayer.tools.io.HttpInterface;
import lavaplayer.track.AudioTrack;
import lavaplayer.track.AudioTrackInfo;
import lavaplayer.track.DelegatedAudioTrack;
import lavaplayer.track.playback.LocalAudioTrackExecutor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.util.List;
import java.util.StringJoiner;

import static lavaplayer.container.Formats.MIME_AUDIO_WEBM;
import static lavaplayer.tools.FriendlyException.Severity.COMMON;

/**
 * Audio track that handles processing Youtube videos as audio tracks.
 */
public class YoutubeAudioTrack extends DelegatedAudioTrack {
    private static final Logger log = LoggerFactory.getLogger(YoutubeAudioTrack.class);

    private final YoutubeItemSourceManager sourceManager;

    /**
     * @param trackInfo     Track info
     * @param sourceManager Source manager which was used to find this track
     */
    public YoutubeAudioTrack(AudioTrackInfo trackInfo, YoutubeItemSourceManager sourceManager) {
        super(trackInfo);

        this.sourceManager = sourceManager;
    }

    private static boolean isBetterFormat(YoutubeTrackFormat format, YoutubeTrackFormat other) {
        YoutubeFormatInfo info = format.getInfo();

        if (info == null) {
            return false;
        } else if (other == null) {
            return true;
        } else if (info.ordinal() != other.getInfo().ordinal()) {
            return info.ordinal() < other.getInfo().ordinal();
        } else {
            return format.getBitrate() > other.getBitrate();
        }
    }

    private static YoutubeTrackFormat findBestSupportedFormat(List<YoutubeTrackFormat> formats) {
        YoutubeTrackFormat bestFormat = null;

        for (YoutubeTrackFormat format : formats) {
            if (isBetterFormat(format, bestFormat)) {
                bestFormat = format;
            }
        }

        if (bestFormat == null) {
            StringJoiner joiner = new StringJoiner(", ");
            formats.forEach(format -> joiner.add(format.getType().toString()));
            throw new IllegalStateException("No supported audio streams available, available types: " + joiner);
        }

        return bestFormat;
    }

    @Override
    public void process(LocalAudioTrackExecutor localExecutor) throws Exception {
        try (HttpInterface httpInterface = sourceManager.getHttpInterface()) {
            FormatWithUrl format = loadBestFormatWithUrl(httpInterface);

            log.debug("Starting track from URL: {}", format.signedUrl);

            if (trackInfo.isStream) {
                processStream(localExecutor, format);
            } else {
                processStatic(localExecutor, httpInterface, format);
            }
        }
    }

    private void processStatic(LocalAudioTrackExecutor localExecutor, HttpInterface httpInterface, FormatWithUrl format) throws Exception {
        try (YoutubePersistentHttpStream stream = new YoutubePersistentHttpStream(httpInterface, format.signedUrl, format.details.getContentLength())) {
            if (format.details.getType().getMimeType().endsWith("/webm")) {
                processDelegate(new MatroskaAudioTrack(trackInfo, stream), localExecutor);
            } else {
                processDelegate(new MpegAudioTrack(trackInfo, stream), localExecutor);
            }
        }
    }

    private void processStream(LocalAudioTrackExecutor localExecutor, FormatWithUrl format) throws Exception {
        if (MIME_AUDIO_WEBM.equals(format.details.getType().getMimeType())) {
            throw new FriendlyException("YouTube WebM streams are currently not supported.", COMMON, null);
        } else {
            try (HttpInterface streamingInterface = sourceManager.getHttpInterface()) {
                processDelegate(new YoutubeMpegStreamAudioTrack(trackInfo, streamingInterface, format.signedUrl), localExecutor);
            }
        }
    }

    private FormatWithUrl loadBestFormatWithUrl(HttpInterface httpInterface) throws Exception {
        YoutubeTrackDetails details = sourceManager.getTrackDetailsLoader()
            .loadDetails(httpInterface, getIdentifier(), true);

        // If the error reason is "Video unavailable" details will return null
        if (details == null) {
            throw new FriendlyException("This video is not available", FriendlyException.Severity.COMMON, null);
        }

        List<YoutubeTrackFormat> formats = details.getFormats(httpInterface, sourceManager.getSignatureResolver());

        YoutubeTrackFormat format = findBestSupportedFormat(formats);

        URI signedUrl = sourceManager.getSignatureResolver()
            .resolveFormatUrl(httpInterface, details.getPlayerScript(), format);

        return new FormatWithUrl(format, signedUrl);
    }

    @Override
    protected AudioTrack makeShallowClone() {
        return new YoutubeAudioTrack(trackInfo, sourceManager);
    }

    @Override
    public ItemSourceManager getSourceManager() {
        return sourceManager;
    }

    private static class FormatWithUrl {
        private final YoutubeTrackFormat details;
        private final URI signedUrl;

        private FormatWithUrl(YoutubeTrackFormat details, URI signedUrl) {
            this.details = details;
            this.signedUrl = signedUrl;
        }
    }
}

File: main/src/main/java/lavaplayer/source/youtube/YoutubeAudioTrack.kt
package lavaplayer.source.youtube

import lavaplayer.container.Formats
import lavaplayer.container.matroska.MatroskaAudioTrack
import lavaplayer.container.mpeg.MpegAudioTrack
import lavaplayer.tools.FriendlyException
import lavaplayer.tools.io.HttpInterface
import lavaplayer.track.AudioTrack
import lavaplayer.track.AudioTrackInfo
import lavaplayer.track.DelegatedAudioTrack
import lavaplayer.track.playback.LocalAudioTrackExecutor
import mu.KotlinLogging
import java.net.URI

/**
 * Audio track that handles processing YouTube videos as audio tracks.
 * @param trackInfo     Track info
 * @param sourceManager Source manager which was used to find this track
 */
class YoutubeAudioTrack(trackInfo: AudioTrackInfo?, override val sourceManager: YoutubeItemSourceManager) :
    DelegatedAudioTrack(trackInfo!!) {
    companion object {
        private val log = KotlinLogging.logger { }

        private fun YoutubeTrackFormat.betterThan(other: YoutubeTrackFormat?): Boolean {
            return when {
                info == null -> false
                other == null -> true
                info.ordinal != other.info.ordinal -> info.ordinal < other.info.ordinal
                else -> bitrate > other.bitrate
            }
        }

        private fun findBestSupportedFormat(formats: List<YoutubeTrackFormat>): YoutubeTrackFormat {
            val bestFormat = formats.fold<YoutubeTrackFormat, YoutubeTrackFormat?>(null) { a, n -> if (n.betterThan(a)) n else a }
            check(bestFormat != null) { "No supported audio streams available, available types: ${formats.joinToString { it.type.mimeType }}" }
            return bestFormat
        }
    }

    override fun makeShallowClone(): AudioTrack =
        YoutubeAudioTrack(info, sourceManager)

    @Throws(Exception::class)
    override fun process(executor: LocalAudioTrackExecutor) {
        sourceManager.httpInterface.use { httpInterface ->
            val format = loadBestFormatWithUrl(httpInterface)
            log.debug { "Starting track from URL: $format.signedUrl{}" }

            if (info.isStream) {
                processStream(executor, format)
            } else {
                processStatic(executor, httpInterface, format)
            }
        }
    }

    @Throws(Exception::class)
    private fun processStatic(localExecutor: LocalAudioTrackExecutor, httpInterface: HttpInterface, format: FormatWithUrl) {
        YoutubePersistentHttpStream(httpInterface, format.signedUrl, format.details.contentLength).use { stream ->
            if (format.details.type.mimeType.endsWith("/webm")) {
                processDelegate(MatroskaAudioTrack(info, stream), localExecutor)
            } else {
                processDelegate(MpegAudioTrack(info, stream), localExecutor)
            }
        }
    }

    @Throws(Exception::class)
    private fun processStream(localExecutor: LocalAudioTrackExecutor, format: FormatWithUrl) {
        if (Formats.MIME_AUDIO_WEBM == format.details.type.mimeType) {
            throw FriendlyException("YouTube WebM streams are currently not supported.", FriendlyException.Severity.COMMON, null)
        }

        sourceManager.httpInterface.use { streamingInterface ->
            val track = YoutubeMpegStreamAudioTrack(info, streamingInterface, format.signedUrl)
            processDelegate(track, localExecutor)
        }
    }

    @Throws(Exception::class)
    private fun loadBestFormatWithUrl(httpInterface: HttpInterface): FormatWithUrl {
        val details = sourceManager.trackDetailsLoader.loadDetails(httpInterface, identifier, true, sourceManager)
            ?: throw FriendlyException("This video is not available", FriendlyException.Severity.COMMON, null)

        // If the error reason is "Video unavailable" details will return null
        val formats = details.getFormats(httpInterface, sourceManager.signatureResolver)
        val format = findBestSupportedFormat(formats)
        val signedUrl = sourceManager.signatureResolver.resolveFormatUrl(httpInterface, details.playerScript, format)

        return FormatWithUrl(format, signedUrl)
    }

    data class FormatWithUrl(val details: YoutubeTrackFormat, val signedUrl: URI)
}
